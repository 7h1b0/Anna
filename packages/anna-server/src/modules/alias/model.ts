import Ajv from 'ajv';
import { v4 as uuidv4 } from 'uuid';
import knex from '../../knexClient';
import aliasSchema from './schema';
import { omit } from '../../utils/utils';

export const TABLE = 'alias';
export const COLUMNS = [
  'aliasId',
  'sceneId',
  'name',
  'description',
  'enabled',
  'startTime',
  'endTime',
  'createdAt',
  'updatedAt',
  'createdBy',
];

type Alias = {
  aliasId: string;
  sceneId: string;
  name: string;
  description?: string;
  startTime?: number;
  endTime?: number;
  enabled: boolean;
  createdAt: number;
  updatedAt: number;
  createdBy: string;
};

export function validate(data: Record<string, unknown>) {
  const ajv = new Ajv();
  return ajv.validate(aliasSchema, data);
}

export async function findAll(): Promise<Alias[]> {
  return knex(TABLE).select(COLUMNS);
}

export async function findById(aliasId: string): Promise<Alias> {
  return knex(TABLE).first(COLUMNS).where('aliasId', aliasId);
}

export async function findByName(name: string): Promise<Alias[]> {
  return knex(TABLE).select(COLUMNS).where('name', name);
}

export async function save({
  sceneId,
  name,
  description,
  createdBy,
  enabled = true,
  startTime,
  endTime,
}: {
  sceneId: string;
  name: string;
  description: string;
  createdBy: string;
  enabled?: boolean;
  startTime?: number;
  endTime?: number;
}): Promise<string> {
  const aliasId = uuidv4();
  await knex(TABLE).insert({
    aliasId,
    sceneId,
    description,
    name,
    enabled,
    createdBy,
    startTime,
    endTime,
  });
  return aliasId;
}

export function remove(aliasId: string) {
  return knex(TABLE).where('aliasId', aliasId).del();
}

export function findByIdAndUpdate(aliasId: string, payload: Partial<Alias>) {
  const safePayload = omit(
    payload,
    'aliasId',
    'createdAt',
    'updatedAt',
    'createdBy',
  );
  return knex(TABLE).update(safePayload).where('aliasId', aliasId);
}

export function isInteger(value: unknown | number): value is number {
  return Number.isInteger(value);
}

/**
 * Given aliases, returns one alias respecting the constraints
 */
export function resolveActiveAlias(aliases: Alias[]): Alias | undefined {
  const time = new Date().getHours();
  const activeAlias = aliases
    .filter((alias) => alias.enabled)
    .filter((alias) => {
      if (isInteger(alias.startTime) && isInteger(alias.endTime)) {
        if (alias.startTime < alias.endTime) {
          return time >= alias.startTime && time < alias.endTime;
        }
        return time >= alias.startTime || time < alias.endTime;
      }
      return true;
    });

  return activeAlias.pop();
}
