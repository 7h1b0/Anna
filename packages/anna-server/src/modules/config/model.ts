import knex from '../../knexClient';

export const TABLE = 'config';

export type Config = {
  isAway: boolean;
};

export async function fetchIsAway(): Promise<{ isAway: true }> {
  return knex(TABLE).first('isAway');
}

export function updateIsAway(isAway: boolean) {
  return knex(TABLE).update({ isAway });
}
