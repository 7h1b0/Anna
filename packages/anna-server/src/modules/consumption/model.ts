import knex from '../../knexClient';

export const TABLE = 'consumption';
export const COLUMNS = ['date', 'value'];

export type Consumption = {
  date: number;
  value: number;
};

export async function findLastEntries(limit: number): Promise<Consumption[]> {
  return knex(TABLE)
    .select(COLUMNS)
    .whereNotNull('value')
    .orderBy('date', 'desc')
    .limit(limit);
}

export async function findLastEntry(): Promise<Consumption> {
  return knex(TABLE)
    .first(COLUMNS)
    .whereNotNull('value')
    .orderBy('date', 'desc')
    .limit(1);
}

export async function save(consumptions: { date: Date; value: number }[]) {
  return knex.transaction((trx) => {
    return Promise.all(
      consumptions.map((consumption) => trx.insert(consumption).into(TABLE)),
    );
  });
}
