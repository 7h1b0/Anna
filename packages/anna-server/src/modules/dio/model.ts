import Ajv from 'ajv';
import knex from '../../knexClient';
import dioSchema from './schema';

export const TABLE = 'dios';

export type Dio = {
  dioId: number;
  roomId: string;
  name: string;
};

export function validate(data: Record<string, unknown>) {
  const ajv = new Ajv();
  return ajv.validate(dioSchema, data);
}

export async function findAll(): Promise<Dio[]> {
  return knex(TABLE).select('dioId', 'roomId', 'name');
}

export async function findById(dioId: number): Promise<Dio> {
  return knex(TABLE).first('dioId', 'roomId', 'name').where('dioId', dioId);
}

export async function findByRoomId(roomId: string): Promise<Dio[]> {
  return knex(TABLE).select('dioId', 'roomId', 'name').where('roomId', roomId);
}

export async function save(newDio: Dio): Promise<Dio> {
  return knex(TABLE)
    .insert(newDio)
    .then(() => newDio);
}

export function remove(dioId: number) {
  return knex(TABLE).where('dioId', dioId).del();
}

export function findByIdAndUpdate(
  dioId: number,
  payload: Partial<Omit<Dio, 'dioId'>>,
) {
  return knex(TABLE).update(payload).where('dioId', dioId);
}
