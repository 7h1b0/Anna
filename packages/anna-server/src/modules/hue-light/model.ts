import knex from '../../knexClient';
export const TABLE = 'lights';
export const COLUMNS = ['roomId', 'lightId'];

export function findRoomId(lightId: number): Promise<string[]> {
  return knex(TABLE).first('roomId').where('lightId', lightId);
}

export function findAll() {
  return knex(TABLE).select(...COLUMNS);
}

export function findByRoomId(roomId: string): Promise<number[]> {
  return knex(TABLE)
    .select('lightId')
    .where('roomId', roomId)
    .then((res) => res.map(({ lightId }) => lightId));
}

export function remove(lightId: number): Promise<number> {
  return knex(TABLE).where('lightId', lightId).del();
}

export async function insertOrUpdate(lightId: number, roomId: string) {
  await knex.raw(
    'INSERT INTO lights (lightId, roomId) values (?, ?) ON DUPLICATE KEY UPDATE roomId=VALUES(roomId)',
    [lightId, roomId],
  );
}
