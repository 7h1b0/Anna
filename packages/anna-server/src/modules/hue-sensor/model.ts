import Ajv from 'ajv';
import knex from '../../knexClient';
import sensorSchema from './schema';

export const TABLE = 'sensors';
export const COLUMNS = ['roomId', 'sensorId'];

export function findByRoomId(roomId: string): Promise<string[]> {
  return knex(TABLE)
    .select('sensorId')
    .where('roomId', roomId)
    .then((res) => res.map(({ sensorId }) => sensorId));
}

export function findAll() {
  return knex(TABLE).select(...COLUMNS);
}

export function remove(sensorId: string): Promise<number> {
  return knex(TABLE).where('sensorId', sensorId).del();
}

export async function insertOrUpdate(sensorId: string, roomId: string) {
  await knex.raw(
    'INSERT INTO sensors (sensorId, roomId) values (?, ?) ON DUPLICATE KEY UPDATE roomId=VALUES(roomId)',
    [sensorId, roomId],
  );
}

export function validate(data: object) {
  const ajv = new Ajv();
  return ajv.validate(sensorSchema, data);
}
