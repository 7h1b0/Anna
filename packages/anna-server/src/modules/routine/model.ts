import Ajv from 'ajv';
import * as parser from 'cron-parser';
import uuidv4 from 'uuid/v4';
import knex from '../../knexClient';
import routineSchema from './schema';
import dispatch from 'utils/dispatch';
import { isBankHoliday, omit } from 'utils/utils';
import * as logger from 'utils/logger';
import { callScene } from 'utils/actions';

export const TABLE = 'routines';
export const COLUMNS = [
  'routineId',
  'name',
  'interval',
  'sceneId',
  'runAtBankHoliday',
  'enabled',
  'createdAt',
  'updatedAt',
  'lastFailedAt',
  'lastRunAt',
  'nextRunAt',
  'failReason',
  'createdBy',
];

export class Routine {
  constructor(
    public routineId: string,
    public sceneId: string,
    public name: string,
    public interval: string,
    public nextRunAt: number | Date,
    public createdBy: string,
    public enabled = true,
    public runAtBankHoliday = true,
    public createdAt?: number | Date,
    public updatedAt?: number | Date,
    public lastFailedAt?: number | Date,
    public lastRunAt?: number | Date,
    public failReason?: string | null,
  ) {}
}

export function computeNextRunAt(cron: string, runAtBankHoliday = true) {
  try {
    const interval = parser.parseExpression(cron, {
      currentDate: new Date(),
    });

    let nextDate = interval.next();
    if (runAtBankHoliday === false) {
      while (isBankHoliday(nextDate.toDate())) {
        nextDate = interval.next();
      }
    }
    return nextDate.toDate();
  } catch (_) {
    logger.error('Impossible to compute nextRunAt');
    return new Date();
  }
}

export function updateAllNextRunAt(routines: Routine[]) {
  return knex.transaction(function(trx) {
    const reqs = routines.map(routine => {
      const nextRunAt = computeNextRunAt(
        routine.interval,
        routine.runAtBankHoliday,
      );

      return knex(TABLE)
        .transacting(trx)
        .update({ nextRunAt })
        .where('routineId', routine.routineId);
    });

    Promise.all(reqs)
      .then(trx.commit)
      .catch(trx.rollback);
  });
}

export function validate(data: object) {
  const ajv = new Ajv();
  return ajv.validate(routineSchema, data);
}

export async function findAll(): Promise<Routine[]> {
  return knex(TABLE).select(COLUMNS);
}

export async function findById(routineId: string): Promise<Routine> {
  return knex(TABLE)
    .first(COLUMNS)
    .where('routineId', routineId);
}

export async function save(
  createdBy: string,
  name: string,
  sceneId: string,
  interval: string,
  enabled?: boolean,
  runAtBankHoliday?: boolean,
) {
  const routineId = uuidv4();
  const nextRunAt = computeNextRunAt(interval, runAtBankHoliday);

  const routine = new Routine(
    routineId,
    sceneId,
    name,
    interval,
    nextRunAt,
    createdBy,
    enabled,
    runAtBankHoliday,
  );

  await knex(TABLE).insert(routine);
  return routine;
}

export async function remove(routineId: string): Promise<number> {
  return knex(TABLE)
    .where('routineId', routineId)
    .del();
}

export async function findByIdAndUpdate(
  routineId: string,
  payload: Routine,
): Promise<number> {
  return knex(TABLE)
    .update(payload)
    .where('routineId', routineId);
}

export async function run(routine: Routine) {
  const done = (error = null) => {
    const failReason = error ? JSON.stringify(error) : null;
    const lastFailedAt = error ? new Date() : routine.lastFailedAt;
    const lastRunAt = new Date();
    const nextRunAt = computeNextRunAt(
      routine.interval,
      routine.runAtBankHoliday,
    );

    const updatedRoutine = {
      ...omit(routine, 'createdAt', 'updatedAt'),
      lastRunAt,
      nextRunAt,
      failReason,
      lastFailedAt,
    };

    return findByIdAndUpdate(routine.routineId, updatedRoutine);
  };

  try {
    await dispatch(callScene(routine.sceneId));
    return done();
  } catch (error) {
    logger.error(JSON.stringify(error));
    return done(error);
  }
}
