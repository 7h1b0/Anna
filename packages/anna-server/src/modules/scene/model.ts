import Ajv from 'ajv';
import uuidv4 from 'uuid/v4';
import knex from '../../knexClient';
import sceneSchema from './schema';
import { TABLE as ACTION_TABLE, findBySceneId } from './action';
import { ToggleDio, ToggleHueLight } from 'utils/actions';

export const TABLE = 'scenes';
export const COLUMNS = [
  'sceneId',
  'name',
  'description',
  'createdAt',
  'updatedAt',
  'createdBy',
];

export class Scene {
  constructor(
    public sceneId: string,
    public name: string,
    public description: string,
    public createdBy: string,
    public createdAt?: number,
    public updatedAt?: number,
  ) {}
}

export class SceneAction extends Scene {
  constructor(
    sceneId: string,
    name: string,
    description: string,
    createdBy: string,
    public actions: (ToggleDio | ToggleHueLight)[],
    createdAt?: number,
    updatedAt?: number,
  ) {
    super(sceneId, name, description, createdBy, createdAt, updatedAt);
  }
}

export function validate(data: object) {
  const ajv = new Ajv();
  return ajv.validate(sceneSchema, data);
}

export async function findAll(): Promise<Scene[]> {
  return knex(TABLE).select(COLUMNS);
}

export async function findById(
  sceneId: string,
): Promise<SceneAction | undefined> {
  const fetchScene = knex(TABLE)
    .first(COLUMNS)
    .where('sceneId', sceneId);
  const fetchActions = findBySceneId(sceneId);

  const [scene, actions] = await Promise.all([fetchScene, fetchActions]);
  if (scene) {
    return { ...scene, actions };
  }
  return;
}

export async function findByIdAndUpdate({
  sceneId,
  name,
  description,
  actions,
}: SceneAction) {
  return knex
    .transaction(async trx => {
      return trx
        .update({ name, description })
        .where('sceneId', sceneId)
        .into(TABLE)
        .then(row => {
          if (row < 1) {
            throw new Error('No scene found');
          }
          return trx
            .del()
            .where('sceneId', sceneId)
            .into(ACTION_TABLE);
        })
        .then(() => {
          return Promise.all(
            actions.map(action => {
              const actionId = uuidv4();
              const formatedAction = {
                type: action.type,
                targetId: action.targetId,
                sceneId,
                actionId,
                body: JSON.stringify(action.body),
              };
              return trx.insert(formatedAction).into(ACTION_TABLE);
            }),
          );
        });
    })
    .catch((err: Error) => {
      if (err.message === 'No scene found') {
        return 0;
      }
      throw err;
    });
}

export async function save({
  name,
  description,
  createdBy,
  actions,
}: SceneAction) {
  const sceneId = uuidv4();
  await knex.transaction(trx => {
    const insertIntoScene = trx
      .insert({
        sceneId,
        description,
        name,
        createdBy,
      })
      .into(TABLE);

    const insertIntoActions = actions.map(action => {
      const formatedAction = {
        type: action.type,
        targetId: action.targetId,
        sceneId,
        body: JSON.stringify(action.body),
        actionId: uuidv4(),
      };
      return trx.insert(formatedAction).into(ACTION_TABLE);
    });

    return Promise.all([insertIntoScene, ...insertIntoActions]);
  });

  return sceneId;
}

export async function remove(sceneId: string): Promise<number[]> {
  return knex.transaction(trx => {
    const removeScene = trx
      .del()
      .where('sceneId', sceneId)
      .into(TABLE);

    const removeActions = trx
      .del()
      .where('sceneId', sceneId)
      .into(ACTION_TABLE);

    return Promise.all([removeScene, removeActions]);
  });
}
