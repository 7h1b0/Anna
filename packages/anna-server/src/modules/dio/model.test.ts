import knex from '../../knexClient';
import * as Dio from './model';

const room1 = '0fc1d78e-fd1c-4717-b610-65d2fa3d01b2';
const room2 = '1fc1d78e-fd1c-4717-b610-65d2fa3d01b2';
const initDios = [
  {
    dioId: 1,
    roomId: room1,
    name: 'test_1',
  },
  {
    dioId: 2,
    roomId: room1,
    name: 'test_2',
  },
  {
    dioId: 3,
    roomId: room2,
    name: 'test_3',
  },
];

describe('Dio', () => {
  beforeEach(async () => {
    await knex(Dio.TABLE).truncate();
    await knex(Dio.TABLE).insert(initDios);
  });

  describe('findAll', () => {
    it('should return all dios', async () => {
      const result = await Dio.findAll();
      expect(result).toEqual(initDios);
    });
  });

  describe('findById', () => {
    it('should return only one dio', async () => {
      const result = await Dio.findById(initDios[0].dioId);
      expect(result).toEqual(initDios[0]);
    });

    it('should return undefined', async () => {
      const result = await Dio.findById(-1);
      expect(result).toBeUndefined();
    });
  });

  describe('save', () => {
    it('should save a new dio', async () => {
      const save = {
        dioId: 4,
        roomId: '0fc1d78e-fd1c-4717-b610-65d2fa3d01b2',
        name: 'test-save',
      };

      const newDio = await Dio.save(save);
      expect(newDio).toEqual(save);
      const dios = await knex(Dio.TABLE).first().where('dioId', save.dioId);
      expect(dios).toEqual(save);
    });

    it('should reject when an id is already taken', async () => {
      const save = {
        dioId: 2,
        roomId: '0fc1d78e-fd1c-4717-b610-65d2fa3d01b2',
        name: 'test-save',
      };

      await expect(Dio.save(save)).rejects.toBeDefined();
    });
  });

  describe('delete', () => {
    it('should delete a dio', async () => {
      await Dio.remove(1);
      const dios = await knex(Dio.TABLE).select();
      expect(dios).toHaveLength(2);
      expect(dios[0]).toHaveProperty('dioId', 2);
      expect(dios[1]).toHaveProperty('dioId', 3);
    });

    it('should not delete a dio', async () => {
      await Dio.remove(-1);
      const dios = await knex(Dio.TABLE).select();
      expect(dios).toHaveLength(initDios.length);
    });
  });

  describe('findByIdAndUpdate', () => {
    it('should update correctly a dio', async () => {
      const rowsAffected = await Dio.findByIdAndUpdate(1, {
        roomId: '0fc1d78e-fd1c-4717-b610-65d2fa3d01b3',
        name: 'updated',
      });
      expect(rowsAffected).toBe(1);
      const dios = await knex(Dio.TABLE).select();
      expect(dios).toMatchSnapshot();
    });

    it('should not update a dio', async () => {
      const rowsAffected = await Dio.findByIdAndUpdate(-1, { name: 'updated' });
      expect(rowsAffected).toBe(0);
      const dios = await knex(Dio.TABLE).select();
      expect(dios).toEqual(initDios);
    });
  });

  describe('validate', () => {
    it('should return true when a dio is valid', () => {
      const dio = {
        dioId: 3,
        roomId: '0fc1d78e-fd1c-4717-b610-65d2fa3d01b2',
        name: 'test-save',
      };

      expect(Dio.validate(dio)).toBeTruthy();
    });

    it('should return false when a dio is missing a props', () => {
      const dio = {
        dioId: 3,
        name: 'test-save',
      };

      expect(Dio.validate(dio)).toBeFalsy();
    });

    it('should return false when a dio has unknow props', () => {
      const dio = {
        dioId: 3,
        roomId: '0fc1d78e-fd1c-4717-b610-65d2fa3d01b2',
        name: 'test-save',
        test: 2,
      };

      expect(Dio.validate(dio)).toBeFalsy();
    });
  });

  describe('findByRoomId', () => {
    it('should return only dios in a given room', async () => {
      const result = await Dio.findByRoomId(room1);
      expect(result).toEqual([initDios[0], initDios[1]]);
    });

    it('should return an empty list when roomId is empty or unknown', async () => {
      const result = await Dio.findByRoomId('hey');
      expect(result).toHaveLength(0);
    });
  });
});
