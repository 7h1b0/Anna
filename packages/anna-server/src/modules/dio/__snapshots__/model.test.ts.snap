// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dio findByIdAndUpdate should update correctly a dio 1`] = `
Array [
  RowDataPacket {
    "dioId": 1,
    "name": "updated",
    "roomId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b3",
  },
  RowDataPacket {
    "dioId": 2,
    "name": "test_2",
    "roomId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
  },
  RowDataPacket {
    "dioId": 3,
    "name": "test_3",
    "roomId": "1fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
  },
]
`;
